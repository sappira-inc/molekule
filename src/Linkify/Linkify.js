import React from 'react';
import PropTypes from 'prop-types';
import { renderToStaticMarkup } from 'react-dom/server';
import ReactMarkdown from 'react-markdown';
import styled from 'styled-components';

const Link = styled.a.attrs({
  target: '_blank',
  rel: 'noopener noreferrer',
})`
  color: inherit;
  text-decoration: underline;
`;

const Root = styled.div`
  div:nth-child(n + 2) {
    margin-top: 5px;
  }
`;

export default function Linkify({ children, source, linkStyle, renderers, ...props }) {
  return (
    <ReactMarkdown
      source={renderToStaticMarkup(children || source)}
      renderers={{
        paragraph: 'div',
        root: p => <Root {...p} />,
        link: p => <Link {...p} style={linkStyle} />,
        ...renderers,
      }}
      {...props}
    />
  );
}

Linkify.propTypes = {
  children: PropTypes.oneOfType([PropTypes.element, PropTypes.string]),
  source: PropTypes.oneOfType([PropTypes.element, PropTypes.string]),
  className: PropTypes.string,
  renderers: PropTypes.shape(),
  linkStyle: PropTypes.shape(),
};

Linkify.defaultProps = {
  className: 're-linkify',
  renderers: {},
};
