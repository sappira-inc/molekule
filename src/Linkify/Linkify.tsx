import React, { FC } from 'react';
import styled from 'styled-components';
import ReactMarkdown from 'react-markdown';

const Link = styled.a.attrs(() => ({
  target: '_blank',
  rel: 'noopener noreferrer',
}))`
  color: inherit;
  text-decoration: underline;
`;

const Root = styled.div`
  white-space: pre-wrap;

  p:first-child {
    margin-top: 0;
  }

  div + div {
    margin-top: 8px;
  }
`;

export interface LinkifyProps {
  source?: string;
  className?: string;
  renderers?: any;
  linkStyle?: any;
}

/** Automatically parse links contained in a body of text. All props except for `linkStyle` are passed through directly to [react-markdown](https://github.com/rexxars/react-markdown), which `Linkify` uses internally. */
export const Linkify: FC<LinkifyProps> = ({ source, linkStyle, renderers, ...props }) => {
  if (typeof source !== 'string') {
    throw new Error('Molekule: source prop must be a valid markdown string');
  }

  // Convert all carriage returns to new lines
  const markdownString = source.replace(/\r/g, '\n');

  return (
    <ReactMarkdown
      source={markdownString}
      renderers={{
        paragraph: 'div',
        root: p => <Root {...p} />,
        link: p => <Link {...p} style={linkStyle} />,
        code: ({ value }) => <div>{value}</div>,
        ...renderers,
      }}
      {...props}
    />
  );
};

Linkify.defaultProps = {
  className: 're-linkify',
  renderers: {},
};
