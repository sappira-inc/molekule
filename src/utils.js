import { get, kebabCase } from 'lodash';
import styled from 'styled-components';

export const themeGet = (theme, lookup, fallback) => get(theme, lookup, fallback);

export const getComponentVariant = (theme, componentName, variant) => {
  const config = themeGet(
    theme,
    `components.${componentName}.variants.${variant}`,
    theme.variants[componentName][variant]
  );
  if (!config) throw new Error(`Molekule: "${variant}" variant not found in theme...`);
  return config;
};

export const getComponentStyle = componentName => ({ theme }) => themeGet(theme, `components.${componentName}.style`);

export const getComponentClassName = ({ className, theme: { classPrefix }, variant }, name) =>
  `${className || ''} ${classPrefix}-${name} ${variant ? `${classPrefix}-${name}-${variant}` : ''}`.trim();

export const createComponent = ({ name, tag = 'div', as, style, props: defaultProps = () => ({}) }) => {
  const component = as ? styled(as) : styled[tag];

  return component.attrs(props => ({
    ...defaultProps(props),
    ...props,
    className: getComponentClassName(props, kebabCase(name)),
  }))`
    ${style}
    ${getComponentStyle(name)}
    ${({ styles = {} }) => styles[name] || {}}
  `;
};
