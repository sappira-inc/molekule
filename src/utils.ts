import kebabCase from 'lodash/kebabCase';
import get from 'lodash/get';
import styled, { CSSProp } from 'styled-components';

export const themeGet = (lookup: any, fallback?: any) => ({ theme }: any = {}) => get(theme, lookup, fallback);

export const getComponentVariant = (theme: any, componentName: string, variant: string) => {
  const config = themeGet(`variants.${componentName}.${variant}`)({
    theme,
  });

  if (!config) throw new Error(`Molekule: "${variant}" variant not found in theme...`);
  return config;
};

export const getComponentSize = (theme: any, componentName: string, size: string) =>
  themeGet(`sizes.${componentName}.${size}`, {})({ theme });

export const getComponentStyle = (componentName: string) => themeGet(`styles.${componentName}`, {});

export const getVariantStyles = (componentName: string, variant: string) =>
  themeGet(`variants.${componentName}.${variant}.style`, {});

interface GetComponentClassNameProps {
  className: string;
  theme: {
    classPrefix: string;
  };
  variant: string;
}

const getComponentClassName = (
  { className, theme: { classPrefix }, variant }: GetComponentClassNameProps,
  name: string
) => `${className || ''} ${classPrefix}-${name} ${variant ? `${classPrefix}-${name}-${variant}` : ''}`.trim();

interface CreateComponentProps<T> {
  name: string;
  tag?: keyof JSX.IntrinsicElements | React.JSXElementConstructor<any>;
  as?: React.ComponentType<any>;
  style?: ((props: any) => CSSProp) | CSSProp;
  props?: (props: T) => any;
}

export const createComponent = <
  T extends object,
  O extends keyof JSX.IntrinsicElements | React.ComponentType<any> = 'div'
>({
  name,
  tag = 'div',
  as,
  style,
  props: getBaseProps = () => ({}),
}: CreateComponentProps<T>) => {
  const component = styled<O>((as || tag) as any);

  return component.attrs<T>((props: any) => {
    const baseProps = getBaseProps(props);
    const finalProps = {
      ...baseProps,
      ...props,
      className: baseProps.className || props.className,
    };

    return {
      ...finalProps,
      className: getComponentClassName(finalProps, kebabCase(name)),
    };
  })<T>`
  ${style}
  ${getComponentStyle(name)}
  ${(p: any) => getVariantStyles(name, p.variant)}
  ${({ styles = {} }) => styles[name] || {}}
`;
};

// eslint-disable-next-line
const focusableFilter = (node: any) =>
  node.tabIndex >= 0 && !node.hasAttribute('disabled') ? NodeFilter.FILTER_ACCEPT : NodeFilter.FILTER_SKIP;

type FilterFn = (node?: any) => any;

const createTreeWalker = (root: any, currentNode: any, filterFn: FilterFn = () => NodeFilter.FILTER_ACCEPT) => {
  const treeWalker = document.createTreeWalker(
    root,
    NodeFilter.SHOW_ELEMENT,
    {
      acceptNode: filterFn,
    },
    false
  );
  treeWalker.currentNode = currentNode;
  return treeWalker;
};

export const findPreviousFocusableElement = (root: any, currentNode: any) => {
  const treeWalker = createTreeWalker(root, currentNode, focusableFilter);

  if (!treeWalker.previousNode() || treeWalker.currentNode === root) {
    return treeWalker.lastChild();
  }

  return treeWalker.currentNode;
};

export const findNextFocusableElement = (root: any, currentNode: any) => {
  const treeWalker = createTreeWalker(root, currentNode, focusableFilter);

  if (!treeWalker.nextNode() || treeWalker.currentNode === root) {
    treeWalker.currentNode = root;
    return treeWalker.firstChild();
  }

  return treeWalker.currentNode;
};

export const getNextCursorPosition = (cursorPos: any, oldValue: any, newValue: any) => {
  const cursorDiff = newValue.length - oldValue.length;
  let nextPosition = cursorPos;

  if (cursorDiff > 1 || cursorDiff < -1) {
    nextPosition += cursorDiff < 0 ? cursorDiff + 1 : cursorDiff - 1;
  }

  return Math.max(nextPosition, 0);
};
