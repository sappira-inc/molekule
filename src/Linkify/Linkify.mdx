---
menu: Core
name: Linkify
---

import { Playground, PropsTable } from 'docz';
import Linkify from './Linkify';

# Linkify

Automatically parse links contained in a body of text. All props except for `linkStyle` are passed through directly to [react-markdown](https://github.com/rexxars/react-markdown), which `Linkify` uses internally.

## Props
<PropsTable of={Linkify} />

## Example

<Playground>
  {() => {
    const markdownString  = `Checkout http://google.com. <span onmouseover="alert('yo')">Another</span> one that's not so great is http://bing.com. You could also checkout duckduckgo.com if you're into that whole privacy thing. <img src="fake.jpg" onError={() => {}} alt="hacker" /> Oh, and you should also check out https://heydoctor.co. And this [markdown link](https://heydoctor.com)\n\n\    I used to be interpreted as a code block but I shouldn't be anymore!!! And strings can have multiple       spaces      inside`;

    return (
      <Linkify linkStyle={{ color: 'magenta' }} source={markdownString} />
    );
  }}
</Playground>
