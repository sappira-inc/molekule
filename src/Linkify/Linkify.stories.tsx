import React from 'react';
import { Linkify } from './Linkify';

export default {
  title: 'Components|Linkify',
  component: Linkify,
};

export const Basic = () => {
  const markdownString = `Checkout http://google.com. <span onmouseover="alert('yo')">Another</span> one that's not so great is http://bing.com. You could also checkout duckduckgo.com if you're into that whole privacy thing. <img src="fake.jpg" onError={() => {}} alt="hacker" /> Oh, and you should also check out https://heydoctor.co. And this [markdown link](https://heydoctor.com)\n\n    I used to be interpreted as a code block but I shouldn't be anymore!!! And strings can have multiple       spaces      inside`;

  return <Linkify linkStyle={{ color: 'magenta' }} source={markdownString} />;
};
