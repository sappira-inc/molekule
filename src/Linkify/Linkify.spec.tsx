import React from 'react';
import { renderWithTheme } from '../test/utils';
import Linkify from './Linkify';

describe('Linkify', () => {
  test('converts links to anchor tags', () => {
    const { asFragment } = renderWithTheme(<Linkify source="Hello! https://google.com is a cool site." />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('escapes HTML entities', () => {
    const { asFragment } = renderWithTheme(
      <Linkify source={`<img src="fake.jpg" onError={() => {}} alt="hacker" /><span>heheh got hacked</span>`} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('can receive linkStyle', () => {
    const { asFragment } = renderWithTheme(
      <Linkify linkStyle={{ color: 'magenta' }} source="Hello! https://google.com is a cool site." />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test('can render markdown', () => {
    const { asFragment } = renderWithTheme(<Linkify source="[this is a link](http://google.com)" />);

    expect(asFragment()).toMatchSnapshot();
  });

  test('reformats code blocks', () => {
    const linkContent = "dear doctor,\n\n    I'm \"5'9\" and 160'";
    const { asFragment } = renderWithTheme(<Linkify source={linkContent} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
