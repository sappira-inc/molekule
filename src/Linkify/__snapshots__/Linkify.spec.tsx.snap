// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Linkify can receive linkStyle 1`] = `
<DocumentFragment>
  .c1 {
  color: inherit;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0 {
  white-space: pre-wrap;
}

.c0 p:first-child {
  margin-top: 0;
}

.c0 div + div {
  margin-top: 8px;
}

<div
    class="c0 re-linkify"
  >
    <div>
      Hello! 
      <a
        class="c1"
        href="https://google.com"
        rel="noopener noreferrer"
        style="color: magenta;"
        target="_blank"
      >
        https://google.com
      </a>
       is a cool site.
    </div>
  </div>
</DocumentFragment>
`;

exports[`Linkify can render markdown 1`] = `
<DocumentFragment>
  .c1 {
  color: inherit;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0 {
  white-space: pre-wrap;
}

.c0 p:first-child {
  margin-top: 0;
}

.c0 div + div {
  margin-top: 8px;
}

<div
    class="c0 re-linkify"
  >
    <div>
      <a
        class="c1"
        href="http://google.com"
        rel="noopener noreferrer"
        target="_blank"
      >
        this is a link
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Linkify converts links to anchor tags 1`] = `
<DocumentFragment>
  .c1 {
  color: inherit;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0 {
  white-space: pre-wrap;
}

.c0 p:first-child {
  margin-top: 0;
}

.c0 div + div {
  margin-top: 8px;
}

<div
    class="c0 re-linkify"
  >
    <div>
      Hello! 
      <a
        class="c1"
        href="https://google.com"
        rel="noopener noreferrer"
        target="_blank"
      >
        https://google.com
      </a>
       is a cool site.
    </div>
  </div>
</DocumentFragment>
`;

exports[`Linkify escapes HTML entities 1`] = `
<DocumentFragment>
  .c0 {
  white-space: pre-wrap;
}

.c0 p:first-child {
  margin-top: 0;
}

.c0 div + div {
  margin-top: 8px;
}

<div
    class="c0 re-linkify"
  >
    <div>
      &lt;img src="fake.jpg" onError={() =&gt; {}} alt="hacker" /&gt;&lt;span&gt;heheh got hacked&lt;/span&gt;
    </div>
  </div>
</DocumentFragment>
`;

exports[`Linkify reformats code blocks 1`] = `
<DocumentFragment>
  .c0 {
  white-space: pre-wrap;
}

.c0 p:first-child {
  margin-top: 0;
}

.c0 div + div {
  margin-top: 8px;
}

<div
    class="c0 re-linkify"
  >
    <div>
      dear doctor,
    </div>
    <div>
      I'm "5'9" and 160'
    </div>
  </div>
</DocumentFragment>
`;
